dhall/src/Dhall/Parser/Expression.hs

```diff
@@ -474,13 +474,12 @@ parsers embedded = Parsers {..}
             alternative04 = (do
                 _openBrace
 
-                whitespace
-
-                _ <- optional (_comma *> whitespace)
+                -- If it fails to parse the _comma we need to recover the
+                -- whitespace that could be prefix for a key
+                _ <- optional $ try (whitespace *> _comma)
 
-                a <- recordTypeOrLiteral
-
-                whitespace
+                src0 <- src whitespace
```

Comment:
This still doesn't seem ideal. I think we want to

1. Parse whitespace
2. Check whether there's a comma
3. If there's a comma:
   a) throw away the whitespace from step 1, and parse more whitespace which we'll pass to `rTOL`
   b) otherwise pass the whitespace from step 1 directly to `rTOL`

I think it would look somewhat like this

```haskell
src0 <- whitespace
mComma <- optional _comma
src1 <- case mComma of
    Just _ -> src whitespace
    Nothing -> return src0
a <- recordTypeOrLiteral src1
```

Reply:
Really clever!

I've tested that change and it works. Since it's not really obvious the reason of doing that for future new-comer developers, I documented it properly.